Test agent execution fails in directory with spaces

**Describe the bug**
Trying to run Native Image tests with the agent fails if the project directory path contains spaces.

(Might also apply when running the application with the agent.)

**To Reproduce**

Steps to reproduce the behavior:
1. Create a directory with a space in its name, for example `path with-spaces`
2. In that directory create a `pom.xml` file with the following content:
    ```xml
    <project xmlns="http://maven.apache.org/POM/4.0.0"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
      <modelVersion>4.0.0</modelVersion>

      <groupId>com.mycompany.app</groupId>
      <artifactId>my-app</artifactId>
      <version>1.0-SNAPSHOT</version>

      <properties>
        <maven.compiler.source>8</maven.compiler.source>
        <maven.compiler.target>8</maven.compiler.target>
      </properties>

      <dependencies>
        <dependency>
          <groupId>org.junit.jupiter</groupId>
          <artifactId>junit-jupiter</artifactId>
          <version>5.10.0</version>
          <scope>test</scope>
        </dependency>
      </dependencies>

      <build>
        <plugins>
          <plugin>
            <groupId>org.graalvm.buildtools</groupId>
            <artifactId>native-maven-plugin</artifactId>
            <version>0.9.24</version>
            <extensions>true</extensions>
            <executions>
              <execution>
                <id>test-native</id>
                <goals>
                  <goal>test</goal>
                </goals>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>
    </project>
    ```
3. Create the test source file `src/test/java/MyTest.java` with the following content:
    ```java
    import static org.junit.jupiter.api.Assertions.assertTrue;

    import org.junit.jupiter.api.Test;

    class MyTest {
        @Test
        void test() {
            assertTrue(true);
        }
    }
    ```
4. Run `mvn clean test -Dagent=true`
  :x: Bug: Surefire execution fails with:
    ```
    Error: Could not find or load main class with-spaces\target\native\agent-output\test\session-{pid}-{datetime}
    Caused by: java.lang.ClassNotFoundException: with-spaces\target\native\agent-output\test\session-{pid}-{datetime}
    ```

**Expected behavior**
Execution does not fail

**Logs**
Surefire log on console:
```
[ERROR] The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was cmd.exe /X /C "C:\Users\...\Downloads\graalvm-jdk-17.0.8+9.1\bin\java -agentlib:native-image-agent=config-output-dir=C:\Users\...\Downloads\test-native\path with-spaces\target\native\agent-output\test\session-{pid}-{datetime} -jar C:\Users\...\AppData\Local\Temp\surefire9367496019182751827\surefirebooter-20230812135354914_3.jar C:\Users\...\AppData\Local\Temp\surefire9367496019182751827 2023-08-12T13-53-54_661-jvmRun1 surefire-20230812135354914_1tmp surefire_0-20230812135354914_2tmp"
```

Note how there are no quotes around the `-agentlib` argument.

**System Info (please complete the following information):**
 - OS: Windows 10
 - GraalVM Version: Graal JDK 17.0.8+9.1 / GraalVM 23.0.1
 - Java Version: 17
 - Plugin version: `native-maven-plugin:0.9.24`

**Additional context**
The Gradle plugin might be affected as well, but I have not tested it.

--- a/native-maven-plugin/docs/functional/tracing-agent.md
+++ b/native-maven-plugin/docs/functional/tracing-agent.md
@@ -23,7 +23,9 @@ Agent output from tests must be stored under `target/native/agent-output/test`; 
 application runs must be stored under `target/native/agent-output/main` unless direct mode changes
 the destination. Application agent runs are attached to the `exec-maven-plugin` execution named by
 the native plugin's `<agentExecutionId>` configuration value; the default execution ID is
-`java-agent` so existing POMs keep working.
+`java-agent` so existing POMs keep working. Generated test-agent arguments derived from project paths
+must remain a single JVM argument when passed through Maven test runners, including when those paths
+contain spaces.
 
 ## 4. Merge and copy
 

--- a/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationWithAgentFunctionalTest.groovy
+++ b/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationWithAgentFunctionalTest.groovy
@@ -97,6 +97,21 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         outputDoesNotContain "containers found"
     }
 
+    @Issue("https://github.com/graalvm/native-build-tools/issues/485")
+    def "test agent works from a project directory with spaces"() {
+        given:
+        withSpacesInProjectDir()
+        withSample("java-application-with-reflection")
+
+        when:
+        // The generated Surefire argLine must keep the test-stage agent output path as one JVM argument. §FS-tracing-agent.3.
+        mvn '-Pnative', '-DquickBuild', 'test', '-Dagent=true', '-DskipNativeTests'
+
+        then:
+        buildSucceeded
+        metadataExistsAt("target/native/agent-output/test/")
+    }
+
     def "test agent with metadata copy task"() {
         given:
         withSample("java-application-with-reflection")

--- a/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationWithAgentFunctionalTest.groovy
+++ b/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationWithAgentFunctionalTest.groovy
@@ -97,6 +97,7 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         outputDoesNotContain "containers found"
     }
 
+    // Verifies agent output paths with spaces remain one JVM argument through Maven test runners. §FS-tracing-agent.3.
     @Issue("https://github.com/graalvm/native-build-tools/issues/485")
     def "test agent works from a project directory with spaces"() {
         given:
@@ -104,7 +105,6 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         withSample("java-application-with-reflection")
 
         when:
-        // The generated Surefire argLine must keep the test-stage agent output path as one JVM argument. §FS-tracing-agent.3.
         mvn '-Pnative', '-DquickBuild', 'test', '-Dagent=true', '-DskipNativeTests'
 
         then:

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeExtension.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeExtension.java
@@ -240,7 +240,7 @@ public class NativeExtension extends AbstractMavenLifecycleParticipant implement
             agent.setValue("agent");
             Xpp3Dom argLine = new Xpp3Dom("argLine");
             // Surefire/Failsafe parse argLine as one command-line string.
-            // Keep spaced output paths as one JVM argument. §FS-tracing-agent.1 §FS-tracing-agent.3.
+            // Keep spaced output paths as one JVM argument. §FS-tracing-agent.3.
             argLine.setValue(quoteAgentArgumentForArgLine(agentArgument));
             configuration.addChild(argLine);
             findOrAppend(configuration, "jvm").setValue(getGraalvmJava());

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeExtension.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeExtension.java
@@ -117,6 +117,19 @@ public class NativeExtension extends AbstractMavenLifecycleParticipant implement
         return "-agentlib:native-image-agent=" + String.join(",", options);
     }
 
+    static String quoteAgentArgumentForArgLine(String agentArgument) {
+        if (!agentArgument.contains(" ")) {
+            return agentArgument;
+        }
+        if (!agentArgument.contains("\"")) {
+            return "\"" + agentArgument + "\"";
+        }
+        if (!agentArgument.contains("'")) {
+            return "'" + agentArgument + "'";
+        }
+        return agentArgument;
+    }
+
     static String agentOutputDirectoryFor(String baseDir, Context context) {
         return (baseDir + "/native/agent-output/" + context).replace('/', File.separatorChar);
     }
@@ -226,7 +239,9 @@ public class NativeExtension extends AbstractMavenLifecycleParticipant implement
             Xpp3Dom agent = findOrAppend(systemPropertyVariables, NATIVEIMAGE_IMAGECODE);
             agent.setValue("agent");
             Xpp3Dom argLine = new Xpp3Dom("argLine");
-            argLine.setValue(agentArgument);
+            // Surefire/Failsafe parse argLine as one command-line string.
+            // Keep spaced output paths as one JVM argument. §FS-tracing-agent.1 §FS-tracing-agent.3.
+            argLine.setValue(quoteAgentArgumentForArgLine(agentArgument));
             configuration.addChild(argLine);
             findOrAppend(configuration, "jvm").setValue(getGraalvmJava());
         });

--- a/native-maven-plugin/src/test/groovy/org/graalvm/buildtools/maven/NativeExtensionTest.groovy
+++ b/native-maven-plugin/src/test/groovy/org/graalvm/buildtools/maven/NativeExtensionTest.groovy
@@ -46,9 +46,11 @@ import org.apache.maven.model.Build
 import org.apache.maven.model.Plugin
 import org.apache.maven.model.PluginExecution
 import org.apache.maven.project.MavenProject
+import org.codehaus.plexus.util.cli.CommandLineUtils
 import org.codehaus.plexus.util.xml.Xpp3Dom
 import spock.lang.Specification
 
+// Protects Maven tracing-agent argLine transport for output paths computed from project directories. §FS-tracing-agent.1 §FS-tracing-agent.3.
 class NativeExtensionTest extends Specification {
     private static final String JUNIT_TRACKING_ENABLED = "junit.platform.listeners.uid.tracking.enabled"
     private static final String JUNIT_TRACKING_OUTPUT_DIR = "junit.platform.listeners.uid.tracking.output.dir"
@@ -84,6 +86,34 @@ class NativeExtensionTest extends Specification {
             assert systemPropertyVariables.getChild(JUNIT_TRACKING_ENABLED).value == "true"
             assert systemPropertyVariables.getChild(JUNIT_TRACKING_OUTPUT_DIR).value == NativeExtension.testIdsDirectory("target")
         }
+    }
+
+    void "test agent argument is quoted for Surefire argLine when output path contains spaces"() {
+        given:
+        def agentArgument = NativeExtension.buildAgentArgument(
+                "/tmp/path with spaces/target",
+                NativeExtension.Context.test,
+                ["config-output-dir={output_dir}"]
+        )
+
+        when:
+        def quotedAgentArgument = NativeExtension.quoteAgentArgumentForArgLine(agentArgument)
+
+        then:
+        quotedAgentArgument == "\"${agentArgument}\""
+        CommandLineUtils.translateCommandline(quotedAgentArgument).toList() == [agentArgument]
+    }
+
+    void "test agent argument is unchanged for Surefire argLine when output path has no spaces"() {
+        given:
+        def agentArgument = NativeExtension.buildAgentArgument(
+                "/tmp/path-without-spaces/target",
+                NativeExtension.Context.test,
+                ["config-output-dir={output_dir}"]
+        )
+
+        expect:
+        NativeExtension.quoteAgentArgumentForArgLine(agentArgument) == agentArgument
     }
 
     private static Plugin plugin(String artifactId) {

--- a/native-maven-plugin/src/test/groovy/org/graalvm/buildtools/maven/NativeExtensionTest.groovy
+++ b/native-maven-plugin/src/test/groovy/org/graalvm/buildtools/maven/NativeExtensionTest.groovy
@@ -50,7 +50,7 @@ import org.codehaus.plexus.util.cli.CommandLineUtils
 import org.codehaus.plexus.util.xml.Xpp3Dom
 import spock.lang.Specification
 
-// Protects Maven tracing-agent argLine transport for output paths computed from project directories. §FS-tracing-agent.1 §FS-tracing-agent.3.
+// Protects Maven tracing-agent argLine transport for output paths computed from project directories. §FS-tracing-agent.3.
 class NativeExtensionTest extends Specification {
     private static final String JUNIT_TRACKING_ENABLED = "junit.platform.listeners.uid.tracking.enabled"
     private static final String JUNIT_TRACKING_OUTPUT_DIR = "junit.platform.listeners.uid.tracking.output.dir"

--- a/native-maven-plugin/src/test/groovy/org/graalvm/buildtools/maven/NativeExtensionTest.groovy
+++ b/native-maven-plugin/src/test/groovy/org/graalvm/buildtools/maven/NativeExtensionTest.groovy
@@ -50,7 +50,6 @@ import org.codehaus.plexus.util.cli.CommandLineUtils
 import org.codehaus.plexus.util.xml.Xpp3Dom
 import spock.lang.Specification
 
-// Protects Maven tracing-agent argLine transport for output paths computed from project directories. §FS-tracing-agent.3.
 class NativeExtensionTest extends Specification {
     private static final String JUNIT_TRACKING_ENABLED = "junit.platform.listeners.uid.tracking.enabled"
     private static final String JUNIT_TRACKING_OUTPUT_DIR = "junit.platform.listeners.uid.tracking.output.dir"
@@ -88,6 +87,7 @@ class NativeExtensionTest extends Specification {
         }
     }
 
+    // Protects Maven tracing-agent argLine transport for output paths computed from project directories. §FS-tracing-agent.3.
     void "test agent argument is quoted for Surefire argLine when output path contains spaces"() {
         given:
         def agentArgument = NativeExtension.buildAgentArgument(

--- a/native-maven-plugin/src/test/groovy/org/graalvm/buildtools/maven/NativeExtensionTest.groovy
+++ b/native-maven-plugin/src/test/groovy/org/graalvm/buildtools/maven/NativeExtensionTest.groovy
@@ -87,7 +87,6 @@ class NativeExtensionTest extends Specification {
         }
     }
 
-    // Protects Maven tracing-agent argLine transport for output paths computed from project directories. §FS-tracing-agent.3.
     void "test agent argument is quoted for Surefire argLine when output path contains spaces"() {
         given:
         def agentArgument = NativeExtension.buildAgentArgument(

